misc/userscripts/add-nextcloud-bookmarks

```diff
@@ -0,0 +1,180 @@
+        )
+    if not ok:
+        message("info", "Dialog box canceled.")
+        exit(0)
+    return text
+
+
+def message(level, text):
+    """display message"""
+    with open(environ["QUTE_FIFO"], "w") as fifo:
+        fifo.write(
+            'message-{} "add-nextcloud-bookmarks userscript: {}"'.format(level, text)
+        )
+        fifo.flush()
+
+
+if "QUTE_FIFO" not in environ:
+    print(
+        "This script is designed to run as a qutebrowser userscript, "
+        "not as a standalone script."
+    )
+    exit(231)
+
+if "QUTE_CONFIG_DIR" not in environ:
+    if "XDG_CONFIG_HOME" in environ:
+        QUTE_CONFIG_DIR = environ["XDG_CONFIG_HOME"] + "/qutebrowser"
+    else:
+        QUTE_CONFIG_DIR = environ["HOME"] + "/.config/qutebrowser"
+else:
+    QUTE_CONFIG_DIR = environ["QUTE_CONFIG_DIR"]
+
+config_file = QUTE_CONFIG_DIR + "/add-nextcloud-bookmarks.ini"
+if path.isfile(config_file):
+    config = configparser.ConfigParser()
+    config.read(config_file)
+    settings = dict(config.items("nextcloud"))
+else:
+    settings = {}
+
+settings_info = [
+    ("host", "host information.", "required"),
+    ("user", "username.", "required"),
+    ("password", "password.", "required"),
+    ("description", "description or leave blank", "optional"),
+    ("tags", "tags (comma separated) or leave blank", "optional"),
+]
+
+# check for settings that need user interaction and clear optional setting if need be
+for setting in settings_info:
+    if setting[0] not in settings:
+        userInput = get_text(setting[0], setting[1])
+        settings[setting[0]] = userInput
+    if setting[2] == "optional":
+        if settings[setting[0]] == "None":
+            settings[setting[0]] = ""
+
+# tags need to be formatted like [], ["one"] or ["one", two"] as a string
+tags = settings["tags"]
+settings["tags"] = str(tags.split(", ")).replace("'", '"')
+
+QUTE_URL = environ["QUTE_URL"]
+api_url = settings["host"] + "/index.php/apps/bookmarks/public/rest/v2/bookmark"
+
+auth = b64encode(format(settings["user"] + ":" + settings["password"]).encode("utf-8"))
+auth = str(auth, "utf-8")
+
+headers = {
+    "Authorization": "Basic {}".format(auth),
+    "Content-Type": "application/json",
+}
+
+# check if there is already a bookmark for the URL
+r = get(
+    "{}?url={}".format(api_url, QUTE_URL),
+    headers=headers,
+    timeout=(3.05, 27),
+)
+if r.status_code != 200:
+    message(
+        "error",
+        "Could not connect to {} with status code {}".format(
+            settings["host"], r.status_code
+        ),
+    )
+    exit(r.status_code)
+
+try:
+    r.json()["data"][0]["id"]
+except IndexError:
+    pass
+else:
+    message("info", "bookmark already exists for {}".format(QUTE_URL))
+    exit(0)
+
+if environ["QUTE_MODE"] == "hints":
+    QUTE_TITLE = QUTE_URL
+else:
+    QUTE_TITLE = environ["QUTE_TITLE"]
+
+# JSON format
+# https://nextcloud-bookmarks.readthedocs.io/en/latest/bookmark.html#create-a-bookmark
+data = (
```

Comment:
It would be cleaner to make this a dict and pass it as the `json` kwarg of `post()`. It wouldn't be necessary to turn the tags into a list in a string after that either